
Cost function is 
\begin{equation}
J(\Theta) = \frac{1}{2m} \sum_{i=1}^{m} (h_\Theta(x^{(i)})-y^{(i)})^2
\end{equation}


Prediction function is
\begin{equation}
h_\Theta(x^{(i)}) = \Theta_0 + \Theta_1 x^{(i)}
\end{equation}

Gradient Descent is
\begin{equation}
\nabla J(\Theta) = (\frac{\delta J}{\delta \Theta_0}, \frac{\delta J}{\delta \Theta_1})
\end{equation}

\begin{equation}
\frac{\delta J}{\delta \Theta_0} = \frac{1}{m} \sum_{i=1}^{m} (h_\Theta(x^{(i)})-y^{(i)})
\end{equation}

\begin{equation}
\frac{\delta J}{\delta \Theta_1} = \frac{1}{m} \sum_{i=1}^{m} (h_\Theta(x^{(i)})-y^{(i)})x^{(i)}
\end{equation}


\begin{equation}
\vec{X} = 
\left[ 
\begin{array}{ccc}
1 & x^{(1)} \\
1 & x^{(2)} \\
1 & x^{(3)} \\
... \\
1 & x^{(m)}
\end{array} 
\right]
\end{equation}


\begin{equation}
\vec{\Theta} = 
\left[
\begin{array}{ccc}
\Theta_0 \\
\Theta_1
\end{array}
\right]
\end{equation}

\begin{equation}
\vec{Y} = 
\left[
\begin{array}{ccc}
y^{(1)} \\
y^{(2)} \\
y^{(3)} \\
... \\
y^{(m)} \\
\end{array}
\right]
\end{equation}


Cost function is 
\begin{equation}
J(\Theta) = \frac{1}{2m} (\vec{X} \vec{\Theta} - \vec{Y})^T (\vec{X} \vec{\Theta} - \vec{Y})
\end{equation}


Gradient Descent is

\begin{equation}
\frac{\delta J}{\delta \Theta_0} = \frac{1}{m} \sum_{i=1}^{m} (h_\Theta(x^{(i)})-y^{(i)})
\end{equation}

\begin{equation}
\frac{\delta J}{\delta \Theta_1} = \frac{1}{m} \sum_{i=1}^{m} (h_\Theta(x^{(i)})-y^{(i)})x^{(i)}
\end{equation}



\begin{equation}
\nabla J(\Theta) = (\frac{\delta J}{\delta \Theta_0}, \frac{\delta J}{\delta \Theta_1}) = 
\frac{1}{m} \vec{X}^T (\vec{X} \vec{\Theta} - \vec{Y})
\end{equation}


In [7]:
import numpy as np

# 一共是20个测试点
m = 20

# x0为都是1的
X0 = np.ones((m, 1), dtype=np.int)
X0

array([[1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1]])

In [5]:
# x1为从1到20的数字
X1 = np.arange(1, m+1).reshape(m, 1)
X1

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12],
       [13],
       [14],
       [15],
       [16],
       [17],
       [18],
       [19],
       [20]])

\begin{equation}
\vec{X} = 
\left[ 
\begin{array}{ccc}
1 & x^{(1)} \\
1 & x^{(2)} \\
1 & x^{(3)} \\
... \\
1 & x^{(m)}
\end{array} 
\right]
\end{equation}

In [8]:
X = np.hstack((X0, X1))
X

array([[ 1,  1],
       [ 1,  2],
       [ 1,  3],
       [ 1,  4],
       [ 1,  5],
       [ 1,  6],
       [ 1,  7],
       [ 1,  8],
       [ 1,  9],
       [ 1, 10],
       [ 1, 11],
       [ 1, 12],
       [ 1, 13],
       [ 1, 14],
       [ 1, 15],
       [ 1, 16],
       [ 1, 17],
       [ 1, 18],
       [ 1, 19],
       [ 1, 20]])

\begin{equation}
\vec{y} = 
\left[
\begin{array}{ccc}
y^{(1)} \\
y^{(2)} \\
y^{(3)} \\
... \\
y^{(m)} \\
\end{array}
\right]
\end{equation}

In [12]:
# y点的值
y = np.array([
    3, 4, 5, 5, 2, 4, 7, 8, 11, 8, 12,
    11, 13, 13, 16, 17, 18, 17, 19, 21
]).reshape(m, 1)
y

array([[ 3],
       [ 4],
       [ 5],
       [ 5],
       [ 2],
       [ 4],
       [ 7],
       [ 8],
       [11],
       [ 8],
       [12],
       [11],
       [13],
       [13],
       [16],
       [17],
       [18],
       [17],
       [19],
       [21]])

In [11]:
# 学习率用 alpha 表示
alpha = 0.01

接下来我们以矩阵向量的形式定义代价函数和代价函数的梯度

Cost function is 
\begin{equation}
J(\Theta) = \frac{1}{2m} (\vec{X} \vec{\Theta} - \vec{y})^T (\vec{X} \vec{\Theta} - \vec{y})
\end{equation}

In [13]:
def error_function(theta, X, y):
    diff = np.dot(X, theta) - y
    return (1./2/m) * np.dot(np.transpose(diff), diff)

\begin{equation}
\nabla J(\Theta) = (\frac{\delta J}{\delta \Theta_0}, \frac{\delta J}{\delta \Theta_1}) = 
\frac{1}{m} \vec{X}^T (\vec{X} \vec{\Theta} - \vec{Y})
\end{equation}

In [14]:
def gradient_function(theta, X, y):
    diff = np.dot(X, theta) - y
    return (1./m) * np.dot(np.transpose(X), diff)

最后就是算法的核心部分，梯度下降迭代计算

\begin{equation}
\vec{\Theta} = 
\left[
\begin{array}{ccc}
\Theta_0 \\
\Theta_1
\end{array}
\right]
\end{equation}

In [16]:
def gradient_descent(X, y, alpha):
    # 给一个初始的值
    theta = np.array([1, 1]).reshape(2, 1)
    gradient = gradient_function(theta, X, y)
    while not np.all(np.absolute(gradient) <= 1e-5):
        theta = theta - alpha * gradient
        gradient = gradient_function(theta, X, y)
    return theta

当梯度小于1e-5时，说明已经进入了比较平滑的状态，类似于山谷的状态，这时候再继续迭代效果也不大了，所以这个时候可以退出循环！

In [19]:
optimal = gradient_descent(X, y, alpha)
print('optimal:', optimal)
print('error function:', error_function(optimal, X, y)[0,0])

optimal: [[0.51583286]
 [0.96992163]]
error function: 1.0149624062331013


Prediction function is
\begin{equation}
h_\Theta(x^{(i)}) = 0.516 + 0.970 x^{(i)}
\end{equation}